### PR TITLE
Tgui ticket system

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Might need a test merge to test disconnects and multi people stuffs

A branch new TGUI based admin ticket system. Cool.

- [x] Implement ticket manager
- [x] Implement ticket messanger
- [x] Add a verb to access the ticket browser easily
- [x] Update from master and remove stupid merge conflicts

image
image

Deadminning closes the window too, we don't want non-admins being able to exploit themselves the ticket window.

## Why It's Good For The Game

Much nicer for admins, includes new features such as claiming and nothing else except looking much better.
Also, it can be moved around to second screens and auto-updates so admins that have a habit of ignoring the bwoinked for 4 minutes will see the unclaimed ticket on their second screen / tab if they have one and can respond.
Claiming can also be used for when you are investigating a ticket but have not answered yet.

## Changelog
:cl:
admin: TGUI Ticket managing system
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
Don't worry about the fucked up changelog and dme file, I'll fix it probably

File: tgui/packages/tgui/interfaces/TicketMessanger.js [hunk 1]
```diff
@@ -0,0 +1,183 @@
+import { multiline, decodeHtmlEntities } from 'common/string';
+import { useBackend, useLocalState } from '../backend';
+import { Box, Button, Flex, Input, Section, Table, Tabs, NoticeBox, Collapsible, BlockQuote, Slider, Divider } from '../components';
+import { Window, Layout } from '../layouts';
+import { round } from 'common/math';
+
+export const TicketMessanger = (props, context) => {
```
Comment (Crossedfall) [inline tgui/packages/tgui/interfaces/TicketMessanger.js]: It's spelled Messenger by the way

**Comment by PowerfulBacon** [comment]
Need to fix runtimes if someone disconnects mid ticket

**Comment by ike709** [comment]
Is this ready for review/testing?

**Comment by PowerfulBacon** [comment]
> Is this ready for review/testing?

Yes

**Comment by ike709** [comment]
There's some minor improvements to be done but overall it works great.

**Comment by PowerfulBacon** [comment]
Test Merge Issues

- [x] Admin PMs show as purple (Only admin to admin PMs show as cyan)
- [x] Claiming a ticket spams message_admins
- [x] Admin to admin tickets result in the claiming going back and forth
- [x] Confirmation if you claim a ticket from someone else
- [ ] Changing mobs stops the ticket panel from updating (UI thing)
- [x] Add a stat panel button
- [x] Show antag status in the ticket

**Comment by ike709** [comment]
And you need to add an explicit "Open Ticket Manager" button to the statpanel for smoothbrains.

**Comment by ike709** [comment]
Another issue: Clicking on a ticket number of a resolved ticket to view the message log changes the "Resolved by" to the person that clicks the ticket number.

File: code/modules/admin/verbs/adminhelp.dm [hunk 2]
```diff
@@ -537,23 +541,33 @@ GLOBAL_DATUM_INIT(ahelp_tickets, /datum/admin_help_tickets, new)
 /datum/admin_help/proc/Claim(key_name = key_name_admin(usr), silent = FALSE)
 	if(claimed_admin == usr)
 		return
+	if(claimed_admin)
+		var/confirm = alert("This ticket is already claimed, overwride claim?",,"Yes", "No")
```
Comment (ike709) [inline code/modules/admin/verbs/adminhelp.dm]:
```suggestion
		var/confirm = alert("This ticket is already claimed, override claim?",,"Yes", "No")

```

File: code/modules/admin/verbs/adminhelp.dm [hunk 3]
```diff
@@ -537,23 +541,33 @@ GLOBAL_DATUM_INIT(ahelp_tickets, /datum/admin_help_tickets, new)
 /datum/admin_help/proc/Claim(key_name = key_name_admin(usr), silent = FALSE)
 	if(claimed_admin == usr)
 		return
+	if(claimed_admin)
+		var/confirm = alert("This ticket is already claimed, overwride claim?",,"Yes", "No")
+		if(confirm == "no")
```
Comment (ike709) [inline code/modules/admin/verbs/adminhelp.dm]:
```suggestion
		if(confirm == "No")

```

File: code/modules/admin/verbs/adminhelp.dm (hunk 2 above)
Comment (PowerfulBacon) [reply code/modules/admin/verbs/adminhelp.dm]: 🤦

**Comment by PowerfulBacon** [comment]
image
Now shows antag status too (because I always forget to check 😳  )

**Comment by PowerfulBacon** [comment]
- [x] Resolving needs to auto claim
- [x] Long ckeys go off the end
- [ ] Flw button on the browswer
- [x] Disconnected players throw a runtime on antag datum
- [x] Disconnected players have no .mob
- [x] Duplication glitch

**Pull request merged**